scroll_derive/examples/derive_custom_ctx.rs

```diff
@@ -0,0 +1,81 @@
+use scroll_derive::{Pread, Pwrite, SizeWith};
+
+/// An example of using a method as the value for a ctx in a derive.
+struct EndianDependent(Endian);
+impl EndianDependent {
+    fn len(&self) -> usize {
+        match self.0 {
+            scroll::Endian::Little => 5,
+            scroll::Endian::Big => 6,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+struct VariableLengthData {
+    buf: Vec<u8>,
+}
+
+impl<'a> TryFromCtx<'a, usize> for VariableLengthData {
+    type Error = scroll::Error;
+
+    fn try_from_ctx(from: &'a [u8], ctx: usize) -> Result<(Self, usize), Self::Error> {
+        let offset = &mut 0;
+        let buf = from.gread_with::<&[u8]>(offset, ctx)?.to_owned();
+        Ok((Self { buf }, *offset))
+    }
+}
+impl<'a> TryIntoCtx<usize> for &'a VariableLengthData {
+    type Error = scroll::Error;
+    fn try_into_ctx(self, dst: &mut [u8], ctx: usize) -> Result<usize, Self::Error> {
+        let offset = &mut 0;
+        for i in 0..(ctx.min(self.buf.len())) {
+            dst.gwrite(self.buf[i], offset)?;
+        }
+        Ok(*offset)
+    }
+}
+impl SizeWith<usize> for VariableLengthData {
+    fn size_with(ctx: &usize) -> usize {
+        *ctx
+    }
+}
+
+#[derive(Debug, PartialEq, Pread, Pwrite, SizeWith)]
+#[repr(C)]
+struct Data {
+    id: u32,
+    timestamp: f64,
+    // You can fix the ctx regardless of what is passed in.
+    #[scroll(ctx = BE)]
+    arr: [u16; 2],
+    // You can use arbitrary expressions for the ctx.
+    // You have access to the `ctx` parameter of the `{pread/gread}_with` inside the expression.
+    // TODO(implement) you have access to previous fields.
+    // TODO(check) will this break structs with fields named `ctx`?.
+    #[scroll(ctx = EndianDependent(ctx.clone()).len())]
```

Comment:
one last thing I'll say: i think the example is actually better if EndianDependent is the ctx for VariableLength data, and inside of it's impl, it calls `len()` to get the read/write size, instead of the ctx == read/write size and relying on `len()` being called/passed in here.

That is clearer and more idiomatic scroll, imho, if there is such a thing :D

However, i understand that you are showing off arbitrary expressions are callable in the ctx here, which is very cool and probably worth retaining as you've written it.

Side note: I'm surprised you have to `ctx.clone()` ? `Endian` is Copy and I'd expect to not need that clone there? I also recall that effectively Ctx's generally need to be copy to be usable within pread/gread but maybe I forgot and that bound was relaxed?

Side Side note: if the ctx does need to be cloned, and you don't clone, how terrible is the error message presented to the user (at derive time)? I imagine it could be pretty bad?

Reply:
```
error[E0308]: mismatched types
  --> examples\derive_custom_ctx.rs:56:36
   |
56 |     #[scroll(ctx = EndianDependent(ctx).len())]
   |                    --------------- ^^^ expected `Endian`, found `&Endian`
   |                    |
   |                    arguments to this struct are incorrect
   |
note: tuple struct defined here
  --> examples\derive_custom_ctx.rs:4:8
   |
4  | struct EndianDependent(Endian);
   |        ^^^^^^^^^^^^^^^
help: consider dereferencing the borrow
   |
56 |     #[scroll(ctx = EndianDependent(*ctx).len())]
```
But if you follow the directions:
```
error[E0614]: type `Endian` cannot be dereferenced
  --> examples\derive_custom_ctx.rs:56:36
   |
56 |     #[scroll(ctx = EndianDependent(*ctx).len())]
   |                                    ^^^^
```
The `clone` works because `Endian.clone()` and `(&Endian).clone` both become `Endian`.
The issue is that the `SizeWith` derive makes this code:
```rs
fn size_with(ctx: &::scroll::Endian)
``` 
but the try_{from,into}_ctx derives make this code:
```rs
fn try_into_ctx(
        self,
        dst: &mut [u8],
        ctx: ::scroll::Endian,
    ) ->
```
and
```rs
fn try_from_ctx(
        src: &'a [u8],
        ctx: ::scroll::Endian,
    ) ->
```

Yeah, its a little hacky :)

If `size_with` was by value then `EndianDependant(ctx)` would work.